api/authorization/namespace_permissions.go

```diff
@@ -24,29 +24,29 @@ type IdentityProvider interface {
 	GetIdentity(context.Context, Info) (Identity, error)
 }
 
-type NamespacePermissions struct {
+type namespacePermissions struct {
 	privilegedClient client.Client
 	identityProvider IdentityProvider
 	rootNamespace    string
 }
 
-func NewNamespacePermissions(privilegedClient client.Client, identityProvider IdentityProvider, rootNamespace string) *NamespacePermissions {
-	return &NamespacePermissions{
+func NewNamespacePermissions(privilegedClient client.Client, identityProvider IdentityProvider, rootNamespace string) *namespacePermissions {
```

Comment:
Returning a private struct is interesting. Maybe we should do more of this.

Reply:
I've been wanting to do this in the handlers as well. Having the type be private means that you can't instantiate it incorrectly outside of the constructor.

Reply:
I have always like this approach, but have avoided it for years, because it used to yield a [golint linter warning](https://github.com/golang/lint/blob/6edffad5e6160f5949cdefc81710b2706fbcd4f6/testdata/unexp-return.go#L11). I tried to find out one more time why exactly returning an unexported type is "annoying" and found [this issue](https://github.com/golang/lint/issues/210). However I am not convinced that it is evil if you return an interface implementation. So I think I agree with you, just putting these references here for additional context.

Reply:
I think I found a valid use case for the annoyance argument. Imagine the following code

```go
package foo

type thing struct {...}

func NewThing() *thing{....}
```

You need to testdrive it, so you create a test package (`foo_test`):

```go
package foo_test
....

var myThing ???? // how do you var the thing to test so that you can use it in `BeforeEach`, etc?

```

Remember that the golang convention is that interfaces are defined wherever needed, not where they are implemented, therefore package `foo` might not have the `Thing` interface at all. 

In this particular case, imagine you had to write unit tests for the `namespacePermissions` struct. How would you do that?

Reply:
In the extreme case, you can create the interface in the test file itself, since that's where it's used. However, in our codebase I think it would be reasonable to use the expected interface type in the test, since that's how the code will actually be used in practice. In this case that would mean `var nsPerm repositories.NamespacePermission`. If that feels awkward since the interface is declared in a different package, then we could move common/shared interfaces into a shared package that contains common interfaces used throughout the codebase.

Personally, I prefer this approach to having to worry about objects being constructed incorrectly.

Reply:
> Personally, I prefer this approach to having to worry about objects being constructed incorrectly.

Absolutely agree! We should be using constructors to create objects and make all objects' field private to guarantee that, there is no argument about it.

What I doubt whether making structs private (i.e. make their names start with a lowercase letter) and returning pointers to those structs be annoying for testing. For example

```go
func NewThing() *Thing
```

vs

```go
func NewThing() *thing
```

> you can create the interface in the test file itself
Feels like a workaround for the sake of testing to me. What about cases when a struct implements a couple of interfaces? Would you need the "productive" interfaces defined wherever they are needed and a test interface that combines them all?

> then we could move the interfaces into a shared package that contains common interfaces used throughout the codebase.
I do not think that this is quite golang idiomatic though.